### PR TITLE
Android WebSockets: include cookie in request

### PR DESCRIPTION
**Pull request by antoinerousseau**
This PR updates #6851 from @srikanthkh, fixing coding conventions and javadoc, and adding a test plan.

## Test plan
Added testing functions into the WebSocketExample page of the UIExplorer, including a tiny http server to set a cookie on demand. Just open the app page and read the instructions :)

### CODE REVIEW

**Comment by facebook-github-bot-0** [comment]
By analyzing the blame information on this pull request, we identified @AndrewJack and @philikon to be potential reviewers.

**Comment by facebook-github-bot-8** [comment]
@antoinerousseau updated the pull request.

**Comment by antoinerousseau** [comment]
cc @zxcpoiu @mkonicek @srikanthkh @satya164 @christopherdro 

missing a test plan (I don't know how to write it)

**Comment by antoinerousseau** [comment]
Also, I merged `facebook:master` into this branch to follow what @srikanthkh did but I'm not sure if you prefer a rebase to . And not sure if I should squash since there are two different authors. Let me know.

**Comment by facebook-github-bot-1** [comment]
@antoinerousseau updated the pull request.

**Comment by zxcpoiu** [comment]
Wow, nice.
About test plan, I think you can refers to https://github.com/facebook/react-native/pull/6889/commits/dcd922b6099592d8a4ef756e9a081561d486dd14
#6889 and #6961 created by @philikon 

He provides WebSocket UIExample, so you may just add a function like sendWithCookies() as a test case

**Comment by zxcpoiu** [comment]
About rebate,
You can squash all merge commit into the main one along with authors.
So you would have 3 commits finally.

1. Original author's commit
2. Your coding style fixing
3. Test case function

File: Examples/UIExplorer/WebSocketExample.js [hunk 1]
```diff
@@ -0,0 +1,325 @@
...
+    }
+    this.state.socket.send(buffer);
+    this.setState({outgoingMessage: ''});
+  };
+
+  _sendHttp = () => {
+    this.setState({
+      fetch: 'fetching',
+    });
+    fetch(this.state.httpUrl).then((response) => {
```
Comment (antoinerousseau) [inline Examples/UIExplorer/WebSocketExample.js]: resolved

File: Examples/UIExplorer/WebSocketExample.js [hunk 2]
```diff
@@ -0,0 +1,325 @@
...
+      if (response.status >= 200 && response.status < 400) {
+        this.setState({
```
Comment (antoinerousseau) [inline Examples/UIExplorer/WebSocketExample.js]: resolved

File: Examples/UIExplorer/WebSocketExample.js [hunk 3]
```diff
@@ -0,0 +1,325 @@
...
+          <Text>To start the HTTP test server:</Text>
+          <Text style={styles.monospace}>
+            ./Examples/UIExplorer/http_test_server.js
+          </Text>
+        </View>
+        <TextInput
+          style={styles.textInput}
+          autoCorrect={false}
+          placeholder="HTTP URL..."
+          onChangeText={(httpUrl) => this.setState({httpUrl})}
```
Comment (antoinerousseau) [inline Examples/UIExplorer/WebSocketExample.js]: resolved

File: Examples/UIExplorer/WebSocketExample.js [hunk 4]
```diff
@@ -0,0 +1,325 @@
...
+          value={this.state.httpUrl}
```
Comment (antoinerousseau) [inline Examples/UIExplorer/WebSocketExample.js]: resolved

File: Examples/UIExplorer/WebSocketExample.js [hunk 5]
```diff
@@ -0,0 +1,325 @@
...
+        />
+        <View style={styles.buttonRow}>
+          <Button
+            onPress={this._sendHttp}
+            label="Send HTTP request to set cookie"
+            disabled={this.state.fetch === 'fetching'}
```
Comment (antoinerousseau) [inline Examples/UIExplorer/WebSocketExample.js]: resolved

File: Examples/UIExplorer/WebSocketExample.js [hunk 6]
```diff
@@ -0,0 +1,325 @@
...
+          />
+        </View>
+        <View style={styles.note}>
+          <Text>
+            {this.state.fetch === 'OK' ? 'Done. Check your WS server console to see if the next WS requests include the cookie (should be "wstest=OK")' : '-'}
```
Comment (antoinerousseau) [inline Examples/UIExplorer/WebSocketExample.js]: resolved

**Comment by antoinerousseau** [comment]
@zxcpoiu should be all good now, tested OK :)

**Comment by zxcpoiu** [comment]
Great job :+1:
LGTM
Now we may wait for others to review and see if it enough to pass.

remember to squash merging commit finally.

**Comment by antoinerousseau** [comment]
the merge commits are squashed. i guess i'll have to merge again and squash again just before the PR gets shipped?

**Comment by zxcpoiu** [comment]
Why I'm still seeing additional merge commit (Merge branch 'master' into android-websocket-cookie) ?

I'm not 100% sure, but as my understanding, it should be have three commits finally in this case. (I have the same confuse before :confused: )

And of course you may be asked to update your PR and rebase again.

Just follow the instructions once reviewers told you.

cc @mkonicek

**Comment by antoinerousseau** [comment]
@zxcpoiu I'm asking because I'm usually used to rebase my PR instead of merging master into it (for cleaner git history), but I've seen other PRs doing it, and I see an automatic Github button saying _Update branch_ that _merges the latest changes from master into this branch._ (see below ⬇️ )
`CONTRIBUTING.md` doesn't say anything about this...

**Comment by zxcpoiu** [comment]
I have exactly same confusion like you.

I did a little bit observe before, looking at some PR which being accepted, found out that some PR get shipped without rebased on top of the latest master commits (it's really hard to always follows the latest commits in such a highly active repo.)

That's why I am **assuming** that it is not necessary to rebase on top of the latest unless there are explicit reasons or conflicts need us to do so.

I guess the principle behind it are simply to keep the commit history clean, mean that each commit corresponds to one intent, one functionality or one fix rather than several commits map to one action. It makes PR easier to review by others.

Hmm, it needs someone who are familiar with that to clarify it a bit then we can help to improve instructions at Contributing.md

**Comment by satya164** [comment]
It doesn't matter how you merge master to your branch. The bot ultimately squashes all commits to a single one before merging.

**Comment by zxcpoiu** [comment]
Thank you @satya164
So

1.

> Squash your commits (git rebase -i).

is basically to provide a clear code commits for people to review, right?

2.
And it is not necessary to keep click `Merge master to your branch` unless there are conflicts or be asked to do so.

is that correct?

**Comment by satya164** [comment]
@zxcpoiu Yes. You would want to merge master if there are conflicts or tests are failing.

**Comment by antoinerousseau** [comment]
these explanations should be added to `CONTRIBUTING.md`

**Comment by antoinerousseau** [comment]
I don't understand the failed CircleCI tests... any clue?

**Comment by mkonicek** [comment]
@antoinerousseau The CirleCI failure looks unrelated. I restarted the build.

**Comment by antoinerousseau** [comment]
@mkonicek it keeps failing :/

can this PR be reviewed/merged anyway?

**Comment by antoinerousseau** [comment]
Could anyone please review this? 😳 😇 :octocat:

**Comment by antoinerousseau** [comment]
why does CircleCI keeps failing? O.o

**Comment by zxcpoiu** [comment]
I'm not sure :{
maybe do a rebase and force push again?

**Comment by antoinerousseau** [comment]
It's sad that this PR got ignored 😢

**Comment by antoinerousseau** [comment]
Fyi I reopened a clean PR: 9114

**Pull request closed without merging**